### PR TITLE
Replace docker with podman in bootstrap image

### PR DESCRIPTION
**Pull request by brianmcarey**
This PR aims to replace docker with podman in the bootstrap image. The bootstrap image is used by a large number of prow jobs in KubeVirt CI including the large e2e jobs.

The changes in this PR are based on #1632 with some minor updates to match changes in podman handling in kubevirtci. 

/cc @dhiller

### CODE REVIEW

File: images/bootstrap/Dockerfile
Comment (dankenigsberg) [inline images/bootstrap/Dockerfile:113]: why is this used instead of the higher-level `dnf -y clean all` ?

File: images/bootstrap/Dockerfile
Comment (dankenigsberg) [inline images/bootstrap/Dockerfile:115]: do we really want to live on the edge ans use `podman` from Fedora testing? I would not add this without a clear explanation why.

File: images/bootstrap/Dockerfile
Comment (dankenigsberg) [inline images/bootstrap/Dockerfile:17]: time has passed, 35 may be better?

Comment (brianmcarey) [reply images/bootstrap/Dockerfile]: yes fedora 35 has been available for a while now - updated.

File: images/bootstrap/Dockerfile
Comment (brianmcarey) [reply images/bootstrap/Dockerfile]: I have tested this without enabling Fedora testing and it works so I removed this.

File: images/bootstrap/Dockerfile
Comment (brianmcarey) [reply images/bootstrap/Dockerfile]: yes makes more sense to use `dnf -y clean all` - updated

File: images/bootstrap/Dockerfile
Comment (brianmcarey) [reply images/bootstrap/Dockerfile]: updated to fedora 36

File: images/bootstrap/Dockerfile
Comment (oshoval) [inline images/bootstrap/Dockerfile:113]: I think that we better auto detect the CRI instead of having the soft link
(kubevirt basically has already auto detect)

File: images/bootstrap/Dockerfile
Comment (oshoval) [inline images/bootstrap/Dockerfile:34]: maybe worth to install `buildah` and `qemu-static-user` 
`buildah` allows multi arch build with podman, and `qemu-static-user` allows the emulation
of cross compiling

File: images/bootstrap/runner.sh
Comment (oshoval) [inline images/bootstrap/runner.sh:99]: 1. timeout until can be used instead please (nice to have)

2. In case of an error and also on success, it just break, shouldn't it exit 1 on error ?
(unless it catch it nicely somewhere else)

**Review by oshoval — COMMENTED** [review]
Very nice thanks

File: github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml
Comment (oshoval) [inline github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml:456]: https right?

File: github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml
Comment (oshoval) [inline github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml:431]: question, do we really need all the parameters ?
what if we can have fewer global ones that if declared the others will be internally declared ?

File: images/bootstrap/Dockerfile
Comment (oshoval) [inline images/bootstrap/Dockerfile:20]: can we use please older version ? for example 3.4 ?
I have this error https://github.com/containers/podman/issues/13327 on CI [1]
and one mentioned WA is to downgrade

[1] https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2125/rehearsal-pull-cluster-network-addons-operator-unit-test/1529420827697614848

Locally i am using podman 3.4.4 and all works good
Thanks

**Comment by oshoval** [comment]
podman is tls secured by default

We have here some TLS failures
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2125/rehearsal-pull-e2e-cluster-network-addons-operator-workflow-k8s/1529420828771356672
"Error: initializing source docker://quay.io/kubevirtci/gocli:2205030954-99bd4d1: pinging container registry quay.io: Get "https://quay.io/v2/": x509: certificate signed by unknown authority"

Maybe we need to apply some certs ? or fix the certs on the image ?
(alternatively less secure solution is to disable podman tls checking, but not sure we want it)

File: images/bootstrap/Dockerfile
Comment (oshoval) [inline images/bootstrap/Dockerfile:32]: It seems we don't have go in the PATH ?

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2130/rehearsal-pull-kubernetes-nmstate-unit-test/1529702244696461312

File: images/bootstrap/Dockerfile
Comment (oshoval) [reply images/bootstrap/Dockerfile:113]: Ok np

File: images/bootstrap/runner.sh
Comment (brianmcarey) [reply images/bootstrap/runner.sh:99]: It should exit 1 instead of break - I will make that update.

File: images/bootstrap/Dockerfile
Comment (brianmcarey) [reply images/bootstrap/Dockerfile:20]: This part of the Dockerfile is used to build podman v4.0.3 

This may no longer needed as there is newer versions of podman available in the fedora repo that wouldn't have been available when the original PR was made. 

I think I will try to run a couple of tests with the version in the fedora 36 repos and we may be able to remove this builder section altogether.

Its podman v4.1.0 in Fedora 36 at the moment  - if that doesn't work we can look at building an older release

File: github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml
Comment (brianmcarey) [reply github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml:431]: yes it might be better to set some of these in the runner.sh script instead.

File: images/bootstrap/runner.sh
Comment (brianmcarey) [reply images/bootstrap/runner.sh:99]: I made that change here https://github.com/kubevirt/project-infra/pull/2140 - a rebase will probably be required to take it in here.

File: images/bootstrap/runner.sh
Comment (0xFelix) [inline images/bootstrap/runner.sh:90]: Maybe it would be better to use the socket for testing if podman works? Like done here: https://github.com/kubevirt/kubevirtci/blob/23ddfc29c03bb1ccaeb79cd204163f8108db1c36/cluster-up/cluster/ephemeral-provider-common.sh#L24

Comment (0xFelix) [reply images/bootstrap/runner.sh:90]: Or here https://github.com/kubevirt/kubevirtci/blob/23ddfc29c03bb1ccaeb79cd204163f8108db1c36/cluster-up/cluster/ephemeral-provider-common.sh#L18

Comment (brianmcarey) [reply images/bootstrap/runner.sh:90]: yes good points - I will look into improving this check - thanks @0xFelix !

Comment (brianmcarey) [reply images/bootstrap/runner.sh:90]: yes good points - I will look into improving this check - thanks @0xFelix !

File: images/bootstrap/Dockerfile
Comment (brianmcarey) [reply images/bootstrap/Dockerfile:82]: I was looking today - the docker command is used quite a bit in the kubevirtci kind providers such as sriov

**Comment by oshoval** [comment]
With the latest image, I see those errors here
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2125/rehearsal-pull-e2e-cluster-network-addons-operator-workflow-k8s/1534095563631366144

```
Enabling IPV6.
/usr/local/bin/runner.sh: line 66: sysctl: command not found
/usr/local/bin/runner.sh: line 67: sysctl: command not found
modprobe: FATAL: Module ip6table_nat not found in directory /lib/modules/4.18.0-365.el8.x86_64
Podman in Container enabled, initializing in podman compatible mode...
```

File: images/bootstrap/runner.sh
Comment (oshoval) [inline images/bootstrap/runner.sh:83]: I think that instead this we would able to use `systemctl enable --now podman.socket` once https://github.com/kubevirt/kubevirtci/pull/824 is merged (because it knows to detect the default root socket path `/run/podman/socket`)

File: images/bootstrap/Dockerfile
Comment (oshoval) [reply images/bootstrap/Dockerfile:82]: About the last comment please see https://github.com/kubevirt/kubevirtci/pull/826

File: images/bootstrap/runner.sh
Comment (oshoval) [reply images/bootstrap/runner.sh:83]: I see thanks

Would it be good idea to include systemd?

What about using the default podman.socket path ? `run/podman/podman.socket`

File: images/bootstrap/runner.sh
Comment (oshoval) [inline images/bootstrap/runner.sh:85]: Btw I tried that with vm based providers, and it didn't work
nor will be needed soon when 
https://github.com/kubevirt/kubevirt/pull/8010
https://github.com/kubevirt/kubevirt/pull/8048
are merged

I think we better fix whatever needed and remove all softlink to docker

Comment (brianmcarey) [reply images/bootstrap/runner.sh:85]: I am not sure if you saw the prowjob linked in the commit - https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_project-infra/2191/rehearsal-check-provision-k8s-1.23/1544729115590922240#1:build-log.txt%3A2152 - its a panic from gocli and I don't see updates in this area in the changes you linked? 

Here is the PR for making gocli work with podman - https://github.com/kubevirt/kubevirtci/pull/554

Comment (brianmcarey) [reply images/bootstrap/runner.sh:85]: It looks to be using a few docker specific modules so it may not be as straight forward as the other scripts.
https://github.com/kubevirt/kubevirtci/blob/111fd507537476c5e5ff9d7a914c6ece8862aaaa/cluster-provision/gocli/docker/docker.go#L15-L17

**Comment by oshoval** [comment]
Wondering, can we have an image that has both docker and podman ?
at least at the beginning
This way it depends on the repo / job itself which to use, according the detection order / env vars

and will be easier to switch between them in case needed, until we stabilize it to use podman

**Comment by oshoval** [comment]
https://github.com/kubevirt/project-infra/pull/2218 was merged

/lgtm

Thanks
(even if there more steps needed we can done later imho)

**Review by dhiller — APPROVED** [review]
Two nits inline, other than that looks great! Thanks!

/approve

/hold in case you want to handle the first one, feel free to remove the hold otherwise

EDIT: i mean the preset name

File: github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml
Comment (brianmcarey) [reply github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml:424]: I think you are right here - it might be better to use the full name so it easier to understand. I will make that update now.

Comment (oshoval) [reply github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml:424]: I think pinc is known a lil
https://github.com/kukaryambik/pinc
imho short name is better

Comment (dhiller) [reply github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml]: @oshoval I disagree. I think we rather should be clear here and not force the reader to have to remember what this preset is about.

And since Brian has made the change already, let's go with it.

**Review by dhiller — APPROVED** [review]
/unhold

File: images/bootstrap/runner.sh
Comment (oshoval) [inline images/bootstrap/runner.sh:45]: here you dont want to change the pinc ?

**Comment by brianmcarey** [comment]
The `build-multiarch-bootstrap-image` prow job failed due to reaching the timeout. I have opened an issue (https://github.com/kubevirt/project-infra/issues/2252) for it and a PR (https://github.com/kubevirt/project-infra/pull/2253) to increase the timeout to 5 hours to unblock these changes.

File: github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml
Comment (oshoval) [reply github/ci/prow-deploy/kustom/base/configs/current/config/config.yaml]: Bttw I think that at the end pinc will come a term like dind, but whatever you prefer

File: images/bootstrap/runner.sh
Comment (oshoval) [inline images/bootstrap/runner.sh:80]: If i run the new image
`docker run -it --privileged  -e PODMAN_IN_CONTAINER_ENABLED=true quay.io/kubevirtci/bootstrap:v20220816-91d8af2 bash`

I see `/podman/podman.sock` is created
because `XDG_RUNTIME_DIR` is empty here

btw it affects kubevirtci / kubevirt, because we look for `/run/podman/podman.sock`
wdyt about using `/run/podman/podman.sock` please ?
`XDG_RUNTIME_DIR` is empty anyhow
this way we won't need to export `KUBEVIRTCI_PODMAN_SOCKET`

Currently because the socket isn't found, it select docker / docker socket which are an alias to podman and its socket

Comment (brianmcarey) [reply images/bootstrap/runner.sh:80]: Yes I think we should update to the path that kubevirtci expects.

Comment (oshoval) [reply images/bootstrap/runner.sh:80]: Not tested yet https://github.com/kubevirt/project-infra/pull/2258
(lets see the presubmit job)

**Pull request merged**
